# ETL PROJECT

In [ ]:
# load imports for data gathering
import json
import pandas as pd # loads pandas library
import requests # loads requests library
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import sqlite3 # sqlalchemy for PostgreSQL
import os

%matplotlib inline

In [ ]:
# function to read the available files from the folder
def process_raw_directory(directory_path):
    """
    Scans the directory and handles each file based on its type.
    """
    directory_path ="C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/raw/"
    
    loaded_data = {}
    sql_scripts = []

    # 1. Iterate through every file in your raw folder
    for filename in os.listdir(directory_path):
        file_path = os.path.join(directory_path, filename)
        
        # Get the file extension 
        ext = os.path.splitext(filename)[1].lower()
        
        try:
            # Handle CSV Data (e.g., twitter_archive_enhanced.csv)
            if ext == '.csv':
                print(f"Loading CSV Data: {filename}")
                loaded_data[filename] = pd.read_csv(file_path)
            
            # Handle TSV Data 
            elif ext == '.tsv':
                print(f"Loading TSV Data: {filename}")
                loaded_data[filename] = pd.read_csv(file_path, sep='\t')
            
            # Handle JSON/TXT Data 
            elif ext in ['.txt', '.json']:
                print(f"Loading JSON Data: {filename}")
                loaded_data[filename] = pd.read_json(file_path, lines=True)
            
            # Handle SQL Files 
            elif ext == '.sql':
                print(f"Detected SQL Script: {filename}")

                with open(file_path, 'r') as f:
                    sql_scripts.append({'filename': filename, 'content': f.read()})
            
            else:
                print(f"Skipping unsupported file type: {filename}")

        except Exception as e:
            print(f"Error processing {filename}: {e}")

    return loaded_data, sql_scripts

# --- Execution ---
raw_folder = 'raw'
data_frames, warehouse_scripts = process_raw_directory(raw_folder)

Loading TSV Data: image_predictions.tsv
Skipping unsupported file type: league_table.xlsx
Detected SQL Script: stolen_vehicles_analysis.sql
Loading JSON Data: tweet_json.txt
Loading CSV Data: twitter_archive_enhanced.csv


In [ ]:
# Function to load datasets
def load_single_file(file_name):
    """
    Professional loader that reads a file from the raw directory 
    by automatically detecting its format from the file extension.
    """
    base_path = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/raw/"
    full_path = os.path.join(base_path, file_name)
    
    # Extract the extension (e.g., '.csv', '.tsv')
    # os.path.splitext returns a tuple: ('filename', '.extension')
    _, extension = os.path.splitext(file_name)
    format_lower = extension.lower()
   
    try:
        if format_lower == '.csv':
            # twitter_archive_enhanced.csv
            return pd.read_csv(full_path)
        
        elif format_lower == '.tsv':
            # image_predictions.tsv
            return pd.read_csv(full_path, sep='\t')
        
        elif format_lower in ['.json', '.txt']:
            # tweet_json.txt
            return pd.read_json(full_path, lines=True)
            
        elif format_lower == '.xlsx':
            return pd.read_excel(full_path, engine='openpyxl')
        
        else:
            print(f"Format '{format_lower}' in file '{file_name}' is not supported yet.")
            return None
            
    except Exception as e:
        print(f"Error loading {file_name}: {e}")
        return None


In [42]:
# 1. Load texts
df_tweets = load_single_file('tweet_json.txt')
df_tweets

created_at                  id              id_str  \
0    2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1    2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2    2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3    2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4    2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   
...                        ...                 ...                 ...   
2349 2015-11-16 00:24:50+00:00  666049248165822465  666049248165822464   
2350 2015-11-16 00:04:52+00:00  666044226329800704  666044226329800704   
2351 2015-11-15 23:21:54+00:00  666033412701032449  666033412701032448   
2352 2015-11-15 23:05:30+00:00  666029285002620928  666029285002620928   
2353 2015-11-15 22:32:08+00:00  666020888022790149  666020888022790144   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2349  Here we have a 1949 1st generation vulpix. Enj...      False   
2350  This is a purebred Piers Morgan. Loves to Netf...      False   
2351  Here is a very happy pup. Big fan of well-main...      False   
2352  This is a western brown Mitsubishi terrier. Up...      False   
2353  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2349           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2349  {'media': [{'id': 666049244999131136, 'id_str'...   
2350  {'media': [{'id': 666044217047650304, 'id_str'...   
2351  {'media': [{'id': 666033409081393153, 'id_str'...   
2352  {'media': [{'id': 666029276303482880, 'id_str'...   
2353  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r..

In [ ]:
# Save txt to postgresql
df_tweets = load_single_file('tweet_json.txt')

if df_tweets is not None:
    cols = ['created_at', 'id', 'id_str', 'full_text', 'retweet_count', 'favorited', 'retweeted']
    
    # Filter columns, drop duplicates, and drop rows with missing IDs
    tweets = df_tweets[cols].drop_duplicates()
    tweets = tweets.dropna(subset=['id'])
    
    print(f"Standardized {len(tweets)} unique tweets.")

    processed_dir = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/"
    
    if not os.path.exists(processed_dir):
        os.makedirs(processed_dir)

    output_file = os.path.join(processed_dir, "processed_tweets.sql")

    # Generate a SQL script with PostgreSQL syntax
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- PostgreSQL Processed Data Table --\n")
        f.write("CREATE TABLE IF NOT EXISTS tweets (\n")
        f.write("    created_at TIMESTAMP,\n")
        f.write("    id BIGINT PRIMARY KEY,\n")
        f.write("    id_str VARCHAR(255),\n")
        f.write("    full_text TEXT,\n")
        f.write("    retweet_count INT,\n")
        f.write("    favorited BOOLEAN,\n")
        f.write("    retweeted BOOLEAN\n")
        f.write(");\n\n")

        # Create Insert Statements for each row
        for _, row in tweets.iterrows():
            # handle single quotes in text to prevent SQL errors
            clean_text = str(row['full_text']).replace("'", "''")
            
            sql = f"INSERT INTO tweets VALUES ('{row['created_at']}', {row['id']}, '{row['id_str']}', " \
                  f"'{clean_text}', {row['retweet_count']}, {row['favorited']}, {row['retweeted']}) " \
                  f"ON CONFLICT (id) DO NOTHING;\n"
            f.write(sql)

    print(f"Professional PostgreSQL script saved to: {output_file}")

Standardized 2354 unique tweets.
Professional PostgreSQL script saved to: C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/processed_tweets.sql


In [ ]:
# 2. Load the csv
df_csv = load_single_file('twitter_archive_enhanced.csv')
df_csv

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [ ]:
# Saving csv to postgresql
df_csv = load_single_file('twitter_archive_enhanced.csv')

if df_csv is not None:
    colsv = ['tweet_id', 'timestamp', 'source', 'text', 'retweeted_status_id', 
             'retweeted_status_user_id', 'retweeted_status_timestamp', 
             'expanded_urls', 'rating_numerator', 'rating_denominator', 
             'name', 'doggo', 'floofer', 'pupper', 'puppo']
    
    # Select columns
    tweet_archive = df_csv[colsv]
    
    tweet_archive = tweet_archive.drop_duplicates(subset=['tweet_id'])
    
    tweet_archive = tweet_archive.dropna(subset=['tweet_id', 'text'])

    print(f"Standardized Archive: {len(tweet_archive)} records prepared.")

    processed_dir = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/"
    output_file = os.path.join(processed_dir, "processed_archive.sql")

    # Generate PostgreSQL-compatible script
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- PostgreSQL Archive Table --\n")
        f.write("CREATE TABLE IF NOT EXISTS twitter_archive (\n")
        f.write("    tweet_id BIGINT PRIMARY KEY,\n")
        f.write("    timestamp TIMESTAMP,\n")
        f.write("    source TEXT,\n")
        f.write("    tweet_text TEXT,\n")
        f.write("    rating_numerator INT,\n")
        f.write("    rating_denominator INT,\n")
        f.write("    dog_stage VARCHAR(50)\n") # Standardizing doggo/pupper/etc
        f.write(");\n\n")

        for _, row in tweet_archive.head(100).iterrows(): # Doing 100 for the demo script
            clean_text = str(row['text']).replace("'", "''")
            sql = f"INSERT INTO twitter_archive (tweet_id, timestamp, tweet_text) " \
                  f"VALUES ({row['tweet_id']}, '{row['timestamp']}', '{clean_text}') " \
                  f"ON CONFLICT (tweet_id) DO NOTHING;\n"
            f.write(sql)

    print(f"Archive SQL script saved to: {output_file}")

Standardized Archive: 2356 records prepared.
Archive SQL script saved to: C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/processed_archive.sql


In [30]:
# 3. Load xlsx
df_xlsx = load_single_file('league_table.xlsx')
df_xlsx

season               league_name               Club  MP   W   D   L  \
0     2008/2009    Belgium Jupiler League     RSC Anderlecht  34  24   5   5   
1     2008/2009    Belgium Jupiler League  Standard de Liège  34  24   5   5   
2     2008/2009    Belgium Jupiler League           KAA Gent  34  17   8   9   
3     2008/2009    Belgium Jupiler League     Club Brugge KV  34  18   5  11   
4     2008/2009    Belgium Jupiler League   SV Zulte-Waregem  34  16   7  11   
...         ...                       ...                ...  ..  ..  ..  ..   
1473  2015/2016  Switzerland Super League            FC Thun  36  10  11  15   
1474  2015/2016  Switzerland Super League      FC St. Gallen  36  10   8  18   
1475  2015/2016  Switzerland Super League           FC Vaduz  36   7  15  14   
1476  2015/2016  Switzerland Super League             Lugano  36   9   8  19   
1477  2015/2016  Switzerland Super League          FC Zurich  36   7  13  16   

      Pts  GF  GA  GD  avg_goals  Pos  
0      77  75  30  45   2.205882    1  
1      77  66  26  40   1.941176    2  
2      59  67  42  25   1.970588    3  
3      59  59  50   9   1.735294    4  
4      55  55  36  19   1.617647    5  
...   ...  ..  ..  ..        ...  ...  
1473   41  45  54  -9   1.250000    6  
1474   38  41  66 -25   1.138889    7  
1475   36  44  60 -16   1.222222    8  
1476   35  46  75 -29   1.277778    9  
1477   34  48  71 -23   1.333333   10  

[1478 rows x 13 columns]

In [69]:
# Save xlsx in postgresql format
df_xlsx = load_single_file('league_table.xlsx')

if df_xlsx is not None:
    colsx = ['Club', 'MP', 'W', 'D', 'L', 'Pts', 'GF', 'GA', 'GD', 'avg_goals', 'Pos']
    
    # Select columns and perform cleaning
    league_table = df_xlsx[colsx].copy()

    # Drop duplicates
    league_table = league_table.drop_duplicates()
    league_table = league_table.dropna(subset=['Club', 'Pos'])
    
    # Ensuring Pos and Pts are integers for the warehouse
    league_table['Pos'] = league_table['Pos'].astype(int)
    league_table['Pts'] = league_table['Pts'].astype(int)

    print(f"Standardized League Table: {len(league_table)} clubs processed.")

    processed_dir = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/"
    output_file = os.path.join(processed_dir, "processed_league_table.sql")

    # PostgreSQL-compatible script
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- PostgreSQL League Table --\n")
        f.write("CREATE TABLE IF NOT EXISTS league_table (\n")
        f.write("    pos INT PRIMARY KEY,\n")
        f.write("    club VARCHAR(255),\n")
        f.write("    mp INT,\n")
        f.write("    w INT,\n")
        f.write("    d INT,\n")
        f.write("    l INT,\n")
        f.write("    pts INT,\n")
        f.write("    gf INT,\n")
        f.write("    ga INT,\n")
        f.write("    gd INT,\n")
        f.write("    avg_goals NUMERIC(10, 2)\n")
        f.write(");\n\n")

        # Create Insert Statements
        for _, row in league_table.iterrows():
            # Handle apostrophes in club names (if applicable)
            clean_club = str(row['Club']).replace("'", "''")
            
            sql = (f"INSERT INTO league_table VALUES ({row['Pos']}, '{clean_club}', "
                   f"{row['MP']}, {row['W']}, {row['D']}, {row['L']}, {row['Pts']}, "
                   f"{row['GF']}, {row['GA']}, {row['GD']}, {row['avg_goals']}) "
                   f"ON CONFLICT (pos) DO UPDATE SET pts = EXCLUDED.pts;\n")
            f.write(sql)

    print(f"League SQL script saved to: {output_file}")

Standardized League Table: 1478 clubs processed.
League SQL script saved to: C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/processed_league_table.sql


In [ ]:
# 4. Load TSV
df_tsv = load_single_file('image_predictions.tsv')
df_tsv

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [ ]:
# Save TSV into postgresql
df_tsv = load_single_file('image_predictions.tsv')

if df_tsv is not None:
    colst = ['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_dog', 'p2', 'p2_dog', 'p3', 'p3_dog']
    
    tsv = df_tsv[colst].copy()
    
    # Remove duplicate predictions based on tweet_id
    tsv = tsv.drop_duplicates(subset=['tweet_id'])
    
    # Drop rows missing the primary key (tweet_id) or the image link
    tsv = tsv.dropna(subset=['tweet_id', 'jpg_url'])
    
    # Standardize text
    for col in ['p1', 'p2', 'p3']:
        tsv[col] = tsv[col].str.replace('_', ' ').str.title()

    print(f"Standardized Image Predictions: {len(tsv)} records processed.")

    processed_dir = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/"
    output_file = os.path.join(processed_dir, "processed_image_predictions.sql")

    # PostgreSQL-compatible script
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- PostgreSQL Image Predictions Table --\n")
        f.write("CREATE TABLE IF NOT EXISTS image_predictions (\n")
        f.write("    tweet_id BIGINT PRIMARY KEY,\n")
        f.write("    jpg_url TEXT,\n")
        f.write("    img_num INT,\n")
        f.write("    p1_breed TEXT,\n")
        f.write("    p1_is_dog BOOLEAN,\n")
        f.write("    p2_breed TEXT,\n")
        f.write("    p2_is_dog BOOLEAN,\n")
        f.write("    p3_breed TEXT,\n")
        f.write("    p3_is_dog BOOLEAN\n")
        f.write(");\n\n")

        # Create Insert Statements
        for _, row in tsv.iterrows():
            # SQL string escaping for breed names
            p1 = str(row['p1']).replace("'", "''")
            p2 = str(row['p2']).replace("'", "''")
            p3 = str(row['p3']).replace("'", "''")
            
            sql = (f"INSERT INTO image_predictions VALUES ({row['tweet_id']}, '{row['jpg_url']}', "
                   f"{row['img_num']}, '{p1}', {row['p1_dog']}, '{p2}', "
                   f"{row['p2_dog']}, '{p3}', {row['p3_dog']}) "
                   f"ON CONFLICT (tweet_id) DO NOTHING;\n")
            f.write(sql)

    print(f"Image Predictions SQL script saved to: {output_file}")

Standardized Image Predictions: 2075 records processed.
Image Predictions SQL script saved to: C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/processed_image_predictions.sql


In [73]:
# Using SQLite to run scripts

def build_warehouse(db_name, script_list):
    # Connect to the database (creates it if it doesn't exist)
    conn = sqlite3.connect(db_name)
    cursor = conn.cursor()
    
    print(f"Connecting to Warehouse: {db_name}")
    
    processed_dir = "C:/Users/Davie/Documents/GitHub/ETL_Data_Warehouse_Project/data/processed/"
    
    for script_file in script_list:
        path = os.path.join(processed_dir, script_file)
        if os.path.exists(path):
            with open(path, 'r', encoding='utf-8') as f:
                sql_script = f.read()
                # executescript handles multiple commands in one go
                cursor.executescript(sql_script)
                print(f"Successfully executed: {script_file}")
    
    conn.commit()
    conn.close()
    print("Data Warehouse Population Complete.")

# Run the automation
scripts_to_run = [
    "processed_tweets.sql", 
    "processed_archive.sql", 
    "processed_league_table.sql", 
    "processed_image_predictions.sql"
]
build_warehouse("My_Data_Warehouse.db", scripts_to_run)

Connecting to Warehouse: My_Data_Warehouse.db
Successfully executed: processed_tweets.sql
Successfully executed: processed_archive.sql
Successfully executed: processed_league_table.sql
Successfully executed: processed_image_predictions.sql
Data Warehouse Population Complete.


In [ ]:
from sqlalchemy import create_engine

# Connection format: postgresql://username:password@localhost:port/database_name
# Default port is 5432
engine = create_engine('postgresql://postgres:YOUR_PASSWORD@localhost:5432/ETL_Warehouse')

# To load your dataframes directly into the new warehouse:
# tweets.to_sql('tweets_table', engine, if_exists='replace', index=False)

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. 

**Quality issues**

- Issues related to the data content (dirty data). We check for four quality diemnsions, completeness, validity, accuracy and consistency.

**Tidiness issues**

- Issues related to the data structure (messy data). We check whether or not each variable forms a column, each observation forms a row or each type of observational unit forms a table

**1. visual assessment**
    - viewing the data without code

In [ ]:
# twitter_archive_enhanced.csv
twitter_archive.head()

- Missing data (NaN, none)
- Non-descriptive columns (source, name, text)
- inconsistent rating denominator
- Extremely low and high rating numerator
- invalid names under name column (a, an, none)

In [ ]:
#'image_predictions.tsv'
path="C:/Users/Davie/Documents/GitHub/data_wrangling/data/"
image_pred=pd.read_csv(path + 'image_predictions.tsv', sep='\t')
image_pred

- None-descriptive column names for the rating algorithms (p's	p's_conf	p's_dog)
- Tidyness issues, p1, p2 and p3 columns 

In [ ]:
#tweet_json.csv
path="C:/Users/Davie/Documents/GitHub/data_wrangling/data/"
tweet_json=pd.read_csv(path + 'tweet_json.csv')
tweet_json

- low retweet_count for tweet_id 886267009285017600

**2. programmatic assessement**

    -checking data issues with (code) python methods.We use .sample() .shape, .describe(), .info(), .dtypes, .nunique()

In [ ]:
twitter_archive.columns # list all columns 

In [ ]:
twitter_archive.shape # assess the dimensions of the data

In [ ]:
twitter_archive.nunique() # assess the number of unique values of each column

In [ ]:
twitter_archive.info() # assesscheck missing and data types of each column

In [ ]:
twitter_archive[twitter_archive.duplicated()] # assess duplicate rows

No duplicates, timestamp column data type should be datetime

In [ ]:
# assess the number of uniques for 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'
twitter_archive[['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']].nunique()

Many retweets that nedd to be removed according to the project motivation

In [ ]:
list_names=[] # empty list
for names in twitter_archive['name']:
    if len(names)<=3: # check if the name has less than 3 characters
        list_names.append(names)
funny_names=pd.Series(list_names).value_counts() # convert to pandas series and check counts per name
funny_names

Some of these names ('a','by','not', 'his', 'an', 'all', 'life', 'the')do not appear as valid names. The following cells contain codes used to examine a few of the them

In [ ]:
twitter_archive_none=twitter_archive[twitter_archive['name']=='None'] # load texts for dogs having 'None' as the names
twitter_archive_none[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

Some of the texts may contain dog names and stages, while others contain None

In [ ]:
twitter_archive_a=twitter_archive[twitter_archive['name']=='a'] # load texts for dogs having 'a' as the names
twitter_archive_a[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'a' is not the correct dog name. Some of the texts contain correct dog names and stages but wrongly extracted, while others contain None

In [ ]:
twitter_archive_an=twitter_archive[twitter_archive['name']=='an'] # load texts for dogs having 'an' as the names
twitter_archive_an[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'an' is not the correct dog name. Though, some of the texts contain correct dog names but wrongly extracted, while others contain None or correct dog stages

In [ ]:
twitter_archive_the=twitter_archive[twitter_archive['name']=='the'] # load texts for dogs having 'the' as the names
twitter_archive_the[['tweet_id','text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'the' is not the correct dog name. However, the text may or may not contain dog name

In [ ]:
twitter_archive_one=twitter_archive[twitter_archive['name']=='one'] # load texts for dogs having 'one' as the names
twitter_archive_one[['tweet_id','text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'one' is not the correct dog name. Texts may not contain dog names. puppers wrongly extracted

In [ ]:
twitter_archive_all=twitter_archive[twitter_archive['name']=='all'] # load texts for dogs having 'all' as the names
twitter_archive_all[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'all' is not the correct dog name

In [ ]:
twitter_archive_not=twitter_archive[twitter_archive['name']=='not'] # load texts for dogs having 'not' as the names
twitter_archive_not[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'not' isn not the correct dog name

In [ ]:
twitter_archive_by=twitter_archive[twitter_archive['name']=='by'] # load texts for dogs having 'by' as the names
twitter_archive_by[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

by is not the correct dog name

In [ ]:
twitter_archive_my=twitter_archive[twitter_archive['name']=='my'] # load texts for dogs having 'my' as the names
twitter_archive_my[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

name is Zoey, wrongly extracted as 'my'

In [ ]:
twitter_archive_old=twitter_archive[twitter_archive['name']=='old'] # load texts for dogs having 'old' as the names
twitter_archive_old[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'old' is not the correct dog name

In [ ]:
twitter_archive_his=twitter_archive[twitter_archive['name']=='his'] # load texts for dogs having 'his' as the names
twitter_archive_his[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

The correct dog name is 'Quizno' and not 'his'

In [ ]:
twitter_archive_just=twitter_archive[twitter_archive['name']=='just'] # load texts for dogs having 'just' as the names
twitter_archive_just[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'just' is not the correct dog name for this group

In [ ]:
twitter_archive_life=twitter_archive[twitter_archive['name']=='life'] # load texts for dogs having 'life' as the names
twitter_archive_life[['text', 'name', 'doggo', 'floofer', 'pupper', 'puppo']]

'life' is not a dog name

**These texts do appear to contain the dog's name**

In [ ]:
image_pred.columns # list all columns

Non-descriptive columns

In [ ]:
image_pred.info() # assess the missing and data types for each column

tweet_id data type is a string

In [ ]:
image_pred.nunique() # assess the unique values in each column

Out of 2075 tweet only 2009 had unique image url

In [ ]:
image_pred[image_pred['jpg_url'].duplicated()].jpg_url.head(10) # assess duplicate image url

66 urls are repeated, pointing to the same image. I want to assess url with id 1315

In [ ]:
image_pred[image_pred['jpg_url']=='https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg'] # assess duplicate rows

The two image url's are the same and show the same image when clicked. The information about the rating algorithm is also the same, except the tweet_id. 

We need a unique tweet-id with a unique image url

In [ ]:
image_pred[image_pred.duplicated()] # assess duplicates

No duplicate tweet_ids

In [ ]:
p1_sorted=image_pred.p1.sort_values() # assess the values under p1-golden retriever
p1_sorted.head(10)

In [ ]:
p2_sorted=image_pred.p2.sort_values() # assess the values under p2-labrador retriever
p2_sorted.head(10)

The two columns contains similar items, they are related

In [ ]:
tweet_json.shape # assess the dimensions of the data

In [ ]:
tweet_json.columns # list the columns in the data

In [ ]:
tweet_json.info() # assess the missing values and data types

In [ ]:
tweet_json.describe() # assess the descriptive statistics for the data

Minimum value is zero in both retweet_count and favorite _count

In [ ]:
tweet_json[tweet_json.duplicated()] # assess duplicate rows

No duplicate observation

In [ ]:
tweet_json.retweet_count.value_counts().tail(10) # assess the counts for each value in column retweet_count

One observation with zero value, did not get a retweet

In [ ]:
tweet_json[tweet_json['retweet_count']==0] # find the observation with value zero in the retweet_count

In [ ]:
tweet_json.favorite_count.value_counts().head(10) # assess the counts for each value in column favorite_count

179 observations have zero values, the tweets had no favourite

In [ ]:
tweet_json[tweet_json['favorite_count']==0].head() # find the observation with value zero in the favorite_count

### Quality issues

1.Missing values in twitter_archive data for name column 'None'

2.Inaccurate values (dog names) in twitter_archive data for name column ('a', 'an', 'all', 'my', 'not', 'the', 'by', 'such', 'his', 'life', 'one', 'old', 'just')

3.181 rows with retweets in twitter archive data, and extreaneous columns need to be removed as per the project motivation

4.Missing values in twitter_archive data for the dog stages column 'None' 

5.The Tweet image prediction, has unique tweet_id but not image url, which is duplicated
 
6.Non-descriptive columns in twitter archive data ('name', 'text')

7.The values in image predictions under columns p1, p2, p3 are uppercase

8.Incorrect data types:'timestamp', 'retweeted_status_timestamp' datatype is of string

9.Incorrect data types for 'tweet_id, retweeted_status_id', retweeted_status_user_id, in_reply_to_status_id, in_reply_to_user_id

10.Inconsistent rating denominator

### Tidiness issues

1.The four columns for doggo, floofer, pupper, and puppo are dog stages, one variable

2.The tweet_id information in all data sets tweet_json.csv, witter archive data and tweet image prediction are related, hence same observational unit.

## Cleaning Data

Thuis aims to improve the quality and tidiness by correcting the inaccuracies, removing the irrelevant columns, renaming columns and replacing missing values, or droping rows with the missing values based on the assessment already done

Cleaning data uses programmatic data cleaning process, in which every issue identified in the assessment section is first defined followed by codng and testing.

In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data
twitter_archive_clean_a=twitter_archive.copy()
image_predictions_clean=image_pred.copy()
tweet_json_clean=tweet_json.copy()

### Issue #1: Missing values in twitter_archive data for name column 'None'

#### Define: Replace the "None" with corectly extracted names or with NaN using .str.extract() method

#### Code

In [ ]:
# select observations where name is not "None"
twitter_archive_clean_b=twitter_archive_clean_a[twitter_archive_clean_a.name !='None']

In [ ]:
# patterns to extract correct names and dog stages
name_pattern='(?:name(?:d)?)\s{1}(?:is\s)?([A-Za-z]+)|(?:This(?:d)?)\s{1}(?:is\s)([A-Za-z]+)|(?:Meet(?:d)?)\s{1}(?:\s)?([A-Za-z]+)|(?:hello(?:d)?)\s{1}(?:to\s)([A-Za-z]+)|(?:call(?:d)?)\s{1}(?:him\s)([A-Za-z]+)'
stage_pattern='(?i)(pupper|doggo|puppo|floofer)'

In [ ]:
twitter_archive_clean_c=twitter_archive_clean_a[twitter_archive_clean_a.name =='None'].copy() # filter observations whose name is None and make a copy
twitter_archive_clean_c['name']=twitter_archive_clean_c['text'].str.extract(name_pattern, expand=True) # extract correct name from text

In [ ]:
twitter_archive_clean_d=twitter_archive_clean_b.append(twitter_archive_clean_c, ignore_index=True) # join the data tables 
twitter_archive_clean_d.head()

#### Test

In [ ]:
twitter_archive_clean_d.name.value_counts()

### Issue #2: Inaccurate values (dog names) in twitter_archive data for name column ('a', 'an', 'all', 'my', 'not', 'the', 'by', 'such', 'his', 'life', 'one', 'old', 'just')

#### Define: Replace ('a', 'an', 'all',  'not', 'the', 'by', 'such', 'life', 'one', 'old', 'just') with NaN using np.NaN, 'his' with 'Quizno' and 'my' with 'Zoey' using replace() method

#### Code

In [ ]:
to_be_replaced=['a', 'an', 'all', 'not', 'the', 'by', 'such', 'life', 'one', 'old', 'just'] # group the values to be replaced
twitter_archive_clean_d['name']=twitter_archive_clean_d['name'].replace(to_be_replaced,np.NaN) # replace the values with NaN
twitter_archive_clean_d.head()

In [ ]:
twitter_archive_clean_d['name']=twitter_archive_clean_d['name'].replace(['my','his'], ['Zoey', 'Quizno']) # correct my and his names with Zoey and Quizno respectively
twitter_archive_clean_d.head()

#### Test

In [ ]:
twitter_archive_clean_d.name.value_counts() # check whether the names replaced still exist in the data

### Issue #3: The four columns for doggo, floofer, pupper, and puppo for twitter_archive data are dog stages, one variable

#### Define: In twitter_arvive data, melt doggo, floofer, pupper, and puppo columns into one column called dog_stage using .melt() method

#### Code

In [ ]:
unmelted_col=['tweet_id', 'timestamp', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 
              'in_reply_to_status_id', 'in_reply_to_user_id','expanded_urls', 'rating_numerator', 
              'rating_denominator', 'name'] # create columns not to be melted
twitter_archive_clean_e=twitter_archive_clean_d.melt(id_vars=unmelted_col, value_vars=['doggo', 'floofer', 'pupper',
       'puppo'], var_name='to_be_removed', value_name='dog_stage') # melt 'doggo', 'floofer', 'pupper','puppo' into dog_stage

In [ ]:
twitter_archive_clean_e.drop('to_be_removed', axis=1, inplace=True) # drop column to be removed

In [ ]:
twitter_archive_clean_e.drop_duplicates(inplace=True) # drops duplicates from the data

#### Test

In [ ]:
len(twitter_archive_clean_e.columns)==len(twitter_archive_clean_e.columns) # should return false

In [ ]:
twitter_archive_clean_e.dog_stage.value_counts() # assess the observations under dog stage column

### Issue #4: Missing values in twitter_archive data for the dog stages column 'None' 

#### Define: Replace None with the correct extracted dog_stage from the text or with NaN

#### Code

In [ ]:
twitter_archive_clean=twitter_archive_clean_e.copy() # make copy
twitter_archive_clean['dog_stage']=twitter_archive_clean['text'].str.extract(stage_pattern, expand=True) # use regex pattern to extract the correct dog stage names

In [ ]:
twitter_archive_clean.dog_stage.value_counts() # assess the new observations under dog stage column

In [ ]:
stage_caps=['Doggo','Floofer', 'PUPPER', 'Puppo', 'DOGGO', 'Pupper'] # create names in caps to be corrected
cor_stage=['doggo', 'floofer', 'pupper', 'puppo', 'doggo', 'pupper'] # correct names
twitter_archive_clean['dog_stage']=twitter_archive_clean['dog_stage'].replace(stage_caps,cor_stage) # replace the names in caps with the correct ones
twitter_archive_clean.head()

#### Test

In [ ]:
twitter_archive_clean.dog_stage.value_counts() # assess the observations under new dog stage column

In [ ]:
twitter_archive_clean.head() # load new twitter_archive_clean data

### Issue #5: Some values in image predictions data under p1, p2 , and p2 columns are in uppercase

#### Define:Make the values of columns p1, p2, p3 in predictions all lowercase.

#### Code

In [ ]:
image_predictions_clean['p1'] = image_predictions_clean['p1'].str.lower() # converts values in p1 to lowercase
image_predictions_clean['p2'] = image_predictions_clean['p2'].str.lower() # converts values in p2 to lowercase
image_predictions_clean['p3'] = image_predictions_clean['p3'].str.lower() # converts values in p3 to lowercase

#### Test

In [ ]:
image_predictions_clean.p1.unique()
image_predictions_clean.p2.unique()
image_predictions_clean.p3.unique()

### Issue #6: The tweet_id information in all data sets tweet_json.csv, witter archive data and tweet image prediction are relerated, hence same observational unit.

#### Define: Merge the three datasets, weet_json.csv, witter archive data and tweet image prediction into one table called twitter_archive_master using merge() method on tweet_id

#### Code

In [ ]:
# first merge twitter_archive_clean to image_predictions_clean
twitter_archive_master_a=image_predictions_clean.merge(twitter_archive_clean, on='tweet_id', how='inner')
twitter_archive_master_a.head()

In [ ]:
# second merge twitter_json_clean to twitter_archive_master_a
twitter_archive_master_b=twitter_archive_master_a.merge(tweet_json_clean, on='tweet_id', how='inner')

In [ ]:
# Increase number of visible columns in a pandas DataFrame to see all the columns in the newly created twitter_archive_master_b
pd.set_option("display.max_columns",25)

In [ ]:
twitter_archive_master_b.head()

#### Test

In [ ]:
twitter_archive_master_b.columns # should return more number of columns

### Issue #7: 181 rows with retweets in twitter archive data, need to be removed as per the project motivation

#### Define: Remove 181 rows with retweets as well as extraneous columns 'in_reply_to_status_id', 'in_reply_to_user_id','source', 'retweeted_status_id', 'retweeted_status_user_id','retweeted_status_timestamp' by filtering them out

#### Code

In [ ]:
# there are 181 retweets found in "retweeted_status_id", "retweeted_status_user_id" and "retweeted_status_timestamp". We keep the rows that are null and remove the retweets.
twitter_archive_master_c = twitter_archive_master_b[twitter_archive_master_b.retweeted_status_id.isnull()]

In [ ]:
# Remove extraneous columns
cols=['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2','p2_conf', 'p2_dog', 'p3', 
      'p3_conf', 'p3_dog', 'timestamp', 'text', 'expanded_urls','rating_numerator', 'rating_denominator', 
      'name', 'dog_stage','retweet_count', 'favorite_count'] # group the columns to be filtered
twitter_archive_master_d=twitter_archive_master_c.filter(cols, axis=1) # filter the required columns
twitter_archive_master_d.head()

#### Test

In [ ]:
twitter_archive_master_d.columns # assess the filtered columns

### Issue #8: The Tweet image prediction, has unique tweet_id but not image url, which is duplicated

#### Define: Remove duplicate jpg_url in tweet image prediction data, to get a unique tweet-id with a unique jpg_url

#### Code

In [ ]:
twitter_archive_master_e=twitter_archive_master_d.drop_duplicates(subset=['jpg_url']) # remove duplicates based on jpg_url column
twitter_archive_master_e.head()

#### Test

In [ ]:
twitter_archive_master_e.shape==twitter_archive_master_d.shape # should return false

### Issue #9: Non-descriptive columns names  ('name', 'text')

#### Define: Rename the column 'name' as 'dog_name'  and 'text' as 'tweets_text' in the twitter_archive_master using .rename() method

#### Code

In [ ]:
# change name to dog name and text to tweets_text
twitter_archive_master=twitter_archive_master_e.copy()
twitter_archive_master.rename(columns={'name': 'dog_name', 'text': 'tweet_text'}, inplace=True)

#### Test

In [ ]:
if twitter_archive_master.columns.any()=='dog_name' or 'tweet_text': # check if any column contains dog_name or tweet_text
    print('Yes') # should return yes

### Issue #10: None-descriptive column names in tweet image prediction data for the rating algorithms ('jpg_url', 'img_num')

#### Define: Rename the column 'jpg_url', and 'img_num', in the twitter_archive_master using .rename() method

#### Code

In [ ]:
rename_dict={'jpg_url':'image_link', 'img_num':'number_of_images'} # create dictionary for old and new names
twitter_archive_master.rename(columns=rename_dict, inplace=True) # rename the coumns

#### Test

In [ ]:
twitter_archive_master.columns # check the columns in the new data table

In [ ]:
twitter_archive_master.head() # load the twitter_archive_master data table

### Issue #11: Incorrect data types for  'timestamp', datatype is of string

#### Define: Convert 'timestamp' data type to datetime

#### Code

In [ ]:
twitter_archive_master['timestamp']=pd.to_datetime(twitter_archive_master['timestamp']) # convert timestamp into datetime

#### Test

In [ ]:
twitter_archive_master['timestamp'].dtypes # assess the data type for timestamp, should return '<M8[ns]'

### Issue #12: Incorrect data types for 'tweet_id, retweeted_status_id', retweeted_status_user_id, in_reply_to_status_id, in_reply_to_user_id

#### Define: Convert 'tweet_id' to string type

#### Code

In [ ]:
twitter_archive_master['tweet_id']=twitter_archive_master['tweet_id'].astype(str) # convert tweet_id to string

In [ ]:
# create order of indexing of the columns
column_names = ['tweet_id', 'timestamp', 'dog_name', 'dog_stage','retweet_count', 'favorite_count', 
                'rating_numerator', 'rating_denominator', 'p1', 'p1_conf','p1_dog', 'p2','p2_conf', 
                'p2_dog', 'p3', 'p3_conf', 'p3_dog','image_link', 'number_of_images','tweet_text', 'expanded_urls']

twitter_archive_master = twitter_archive_master.reindex(columns=column_names) # reorder the columns

In [ ]:
twitter_archive_master.head()

In [ ]:
twitter_archive_master.drop_duplicates(inplace=True) # remove duplicates

#### Test

In [ ]:
twitter_archive_master['tweet_id'].dtypes # should return O, object

In [ ]:
twitter_archive_master.shape # assess the shape of twitter_archive_master

## Storing Data

Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [ ]:
twitter_archive_master.to_csv('twitter_archive_master.csv', index=False) # store twitter_archive_master dataframe to a csv file named twitter_archive_master.csv

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

In [ ]:
twitter_archive_master=pd.read_csv('twitter_archive_master.csv') # load twitter_archive_master.csv into pandas dataframe
twitter_archive_master.head()

### Insights:
1.Pupper dog stage is the most popular dog stage amongst WeRateDogs’s tweets, favorite and retweets counts. The second, most popular dog stage based on the retweets and favorite counts is doggo. 

2.There is strong linear relationship between the Favourites count and the Retweet, though most of the data is accumulated at the start. This relationship is the same in every dog stage. Also, the distribution for p1, p2 and p3 is really skewed

3.Golden Retriever is the most popular dog breed amongst WeRateDogs’s tweets in terms of the number of image predictions having 139 dogs. The second most popular dog breed is Labrador Retriever also having 95 dogs. Therefore, golden retriever, labrador retriever, pembroke, Chihuahua and pug make top 5 most popular dog breeds

### Visualization

In [ ]:
values=twitter_archive_master.groupby(['dog_stage']).retweet_count.sum().sort_values()
#twitter_archive_master.dog_stage.value_counts()
labels=['floofer', 'puppo', 'doggo', 'pupper']

plt.rcParams['font.size'] = '16'
fig, ax=plt.subplots(figsize=[10,14])
fig.patch.set_facecolor('white')  # Set figure background to white

#explode = (0, 0, 0.2, 0.1)
plt.pie(values, labels=labels, counterclock=False, autopct='%1.1f%%') #explode=explode,  shadow=True, 
plt.title('Dog Stage Popularity Chart-Retweet Counts', fontsize=20)
plt.legend(labels, loc=2)

# Save the plot as a JPG file
plt.savefig('dog_stage_popularity_chart.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

The retweet counts pie chart shows that the most popular dog stage is pupper, which is has 42.5 percent populaity. The second is doggo, followed by puppo and lastly floofer, whose popularity is 2.8 percent

In [ ]:
values=twitter_archive_master.groupby(['dog_stage']).favorite_count.sum().sort_values()
#twitter_archive_master.dog_stage.value_counts()
labels=['floofer', 'puppo', 'doggo', 'pupper']

plt.rcParams['font.size'] = '12'
fig, ax=plt.subplots(figsize=[10,14])
fig.patch.set_facecolor('white')  # Set figure background to white

#explode = (0, 0.1, 0.2, 0.1)
plt.pie(values, labels=labels, counterclock=False, autopct='%1.1f%%') #explode=explode,
plt.title('Dog Stage Popularity Chart based on Favorite Counts', fontsize=20)
plt.legend(labels, loc=4)

# Save the plot as a JPG file
plt.savefig('dog_stage_popularity_chartFa.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

In this case, favorite counts pie chart still shows that the most popular dog stage is pupper with popularity at 43.8 percent, which 1.3 percent increase from the previous pie chart. Doggo is still second most popular dog stage, the percentage has reduced from 39.3 percent for retweet counts to 36.1 percent for favourite counts.

In [ ]:
twitter_archive_master.groupby(['dog_stage']).favorite_count.sum().sort_values()

From the above output, pupper dog stage has the highest sum of of favorite counts

In [ ]:
# plot scatter plot for Retweet Counnts vs favorite Counnts
x=twitter_archive_master.retweet_count
y=twitter_archive_master.favorite_count

fig, ax=plt.subplots(figsize=[10, 6])
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

plt.scatter(x,y,color='blue')
plt.title('Scatter plot for Retweet Counnts vs Favorite Counnts')
plt.xlabel('Retweet Counnts')
plt.ylabel('Favorite Counnts')

# Save the plot as a JPG file
plt.savefig('scatter_R_and_F.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

There is strong linear correlation between retweet counnts and favorite counnts

In [ ]:
# plot scatter plots for retweet counnts and favorite counnts for each dog stage
x1=twitter_archive_master[twitter_archive_master.dog_stage=='pupper'].retweet_count
y1=twitter_archive_master[twitter_archive_master.dog_stage=='pupper'].favorite_count

x2=twitter_archive_master[twitter_archive_master.dog_stage=='doggo'].retweet_count
y2=twitter_archive_master[twitter_archive_master.dog_stage=='doggo'].favorite_count

x3=twitter_archive_master[twitter_archive_master.dog_stage=='puppo'].retweet_count
y3=twitter_archive_master[twitter_archive_master.dog_stage=='puppo'].favorite_count

x4=twitter_archive_master[twitter_archive_master.dog_stage=='floofer'].retweet_count
y4=twitter_archive_master[twitter_archive_master.dog_stage=='floofer'].favorite_count

plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams["figure.autolayout"] = False
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

labels='pupper'
plt.scatter(x1,y1, color='green')
plt.title('Scatter plot for Pupper')
plt.xlabel('Retweet Counnts')
plt.ylabel('Favorite Counnts')
plt.legend([labels], loc=0)

plt.savefig('scatter_for_pupper.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

In [ ]:
plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams["figure.autolayout"] = False
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

labels='doggo'
plt.scatter(x2,y2,color='blue')
plt.title('Scatter plot for Doggo')
plt.xlabel('Retweet Counnts')
plt.ylabel('Favorite Counnts')
plt.legend([labels], loc=0)

plt.savefig('scatter_for_doggo.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

In [ ]:
plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams["figure.autolayout"] = False
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

labels='puppo'
plt.scatter(x3,y3,color='orange')
plt.title('Scatter plot for Puppo')
plt.xlabel('Retweet Counnts')
plt.ylabel('Favorite Counnts')
plt.legend([labels], loc=0)

plt.savefig('scatter_for_puppo.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

In [ ]:
plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams["figure.autolayout"] = False
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

labels='floofer'
plt.scatter(x4,y4,color='black')
plt.title('Scatter plot for Floofer')
plt.xlabel('Retweet Counnts')
plt.ylabel('Favorite Counnts')
plt.legend([labels], loc=0)

plt.savefig('scatter_for_floofer.jpg', format='jpg', dpi=300, bbox_inches='tight')

plt.show()

From the output, there is strong linear correlation between retweet counnts and favorite counnts for each dog stage

In [ ]:
# plot scatter matrix
histogram=twitter_archive_master[['p1_conf','p2_conf','p3_conf']]

plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams['font.size'] = '14' # Set general font size
fig.patch.set_facecolor('white')  # Set figure background to white

pd.plotting.scatter_matrix(histogram, color='black', alpha=0.7)
plt.xticks(rotation=0)

plt.savefig('scatter_matrix.jpg', format='jpg', dpi=300, bbox_inches='tight')
plt.show()

The distributions are all skewed. p1 is left skewed, while p2 and p3 are right skewed

Next, I want to get the most common dog breed

In [ ]:
twitter_archive_breed_type=twitter_archive_master.copy() # make copy of twitter_archive_master

In [ ]:
# put p1, p2 and p3 into one group known as breed type using melt method
twitter_archive_breed_type=twitter_archive_breed_type.melt(id_vars=['tweet_id', 'timestamp', 'dog_name', 'dog_stage','retweet_count', 'favorite_count', 
                 'rating_numerator', 'rating_denominator','p1_conf','p1_dog','p2_conf', 
                 'p2_dog', 'p3_conf', 'p3_dog','image_link', 'number_of_images',
                     'tweet_text', 'expanded_urls'], 
            value_vars=['p1', 'p2', 'p3'], 
            var_name='to_be_removed1', value_name='breed_type')
twitter_archive_breed_type.drop('to_be_removed1', axis=1, inplace=True)

In [ ]:
twitter_archive_breed_type.drop_duplicates(inplace=True) # remove duplicates

In [ ]:
twitter_archive_breed=twitter_archive_breed_type.copy() # make copy

In [ ]:
# put 'p1_dog', 'p2_dog', 'p3_dog' into one column called breed using melt method
twitter_archive_breed=twitter_archive_breed_type.melt(id_vars=['tweet_id', 'timestamp', 'dog_name', 'dog_stage','retweet_count', 'favorite_count', 
                 'rating_numerator', 'rating_denominator','p1_conf','p2_conf', 'p3_conf', 
                                                          'image_link', 'number_of_images',
                     'tweet_text', 'expanded_urls', 'breed_type'], 
            value_vars=['p1_dog', 'p2_dog', 'p3_dog'], 
            var_name='to_be_removed2', value_name='breed')
twitter_archive_breed.drop('to_be_removed2', axis=1, inplace=True)

In [ ]:
twitter_archive_breed.drop_duplicates(subset='tweet_id',inplace=True) # remove duplicates if any

In [ ]:
# select only those which are dog breed
breed=twitter_archive_breed[twitter_archive_breed.breed==True].breed_type.value_counts()

In [ ]:
breed

In [ ]:
color=['green', 'red', 'yellow', 'blue', 'black', 'orange', 'violet', 'brown', 'indigo', 'pink']
breed[breed>27].reset_index()
#reindex(['breed type', 'count'])


In [ ]:
plt.rcParams["figure.figsize"] = [10, 6]
fig.patch.set_facecolor('white')  # Set figure background to white

# Filter and reset index
filtered_breed = breed[breed > 27].reset_index()
filtered_breed.columns = ['index', 'breed_type']  # Ensure proper column names

# Create the barplot
catplot = sb.catplot(
    data=filtered_breed, 
    kind='bar', 
    y='index', 
    x='breed_type', 
    palette='viridis', 
    height=6,  # Adjust height of the plot
    aspect=1.5  # Adjust aspect ratio
)

# Add title and labels
catplot.ax.set_title('Top 10 Dog Breeds', fontsize=16, fontweight='bold')
catplot.ax.set_xlabel('Frequency', fontsize=12)
catplot.ax.set_ylabel('Dog Breed', fontsize=12)

plt.savefig('top10dogs.jpg', format='jpg', dpi=300, bbox_inches='tight')

# Display the plot
plt.show()


The most common dog breed is golden retriever, interms of the number of image predictions. The second most popular dog breed is Labrador Retriever, followed by Pembroke and finaly by Chihuahua